### PR TITLE
provider: Add session persistence

### PR DESCRIPTION
**Pull request by vancluever**
This introduces session persistence to the provider.

This is actually being driven by our needs in getting the provider up and running in our CI for nightly acceptance tests. Unfortunately, the large amount of new TF contexts that the acceptance test suite will make for a single test means that running all of these tests - even in succession versus concurrently - consumes all of the available sessions within vSphere - and that limit is 1000. :fearful:

Introducing this has removed session-related errors from the acceptance tests when running them in CI.

The method has been largely adapted from [govc](https://github.com/vmware/govmomi/tree/master/govc) and the SOAP client session caching has been designed to be compatible with default behaviour in govc - it should actually be able to re-use govc sessions if they exist. However, while it is on by default in govc, it is not in Terraform.

This work has also produced https://github.com/vmware/vic/pull/7464 which will have to be re-vendored once it's merged. For now we are working off a fork.

### CODE REVIEW

**Comment by paultyng** [comment]
I wonder if by default this should use `os.TempDir` instead of home directory by default?

**Comment by paultyng** [comment]
Oh nm, without being a shared location it eliminates the benefit :(

File: vsphere/config_test.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+	vimSessionDir, err := ioutil.TempDir("", "tf-vsphere-test-vimsessiondir")
+	if err != nil {
+		t.Fatalf("error creating VIM session temp directory: %s", err)
+	}
+	restSessionDir, err := ioutil.TempDir("", "tf-vsphere-test-restsessiondir")
+	if err != nil {
+		t.Fatalf("error creating REST session temp directory: %s", err)
+	}
+	defer func() {
+		if err = os.RemoveAll(vimSessionDir); err != nil {
```
Comment (paultyng) [inline vsphere/config_test.go]: split in to 2 defers?

File: vsphere/config_test.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+func TestAccClient_noPersistence(t *testing.T) {
+	vimSessionDir, err := ioutil.TempDir("", "tf-vsphere-test-vimsessiondir")
+	if err != nil {
+		t.Fatalf("error creating VIM session temp directory: %s", err)
+	}
+	restSessionDir, err := ioutil.TempDir("", "tf-vsphere-test-restsessiondir")
+	if err != nil {
+		t.Fatalf("error creating REST session temp directory: %s", err)
+	}
+	defer func() {
```
[the remark this answers is not in the document]
Comment (vancluever) [reply vsphere/config_test.go]: Yeah, good catch. I think I was thinking "well defer will happen even on fatal" but that won't happen for the inner fatals, will it ;p Will fix.

**Comment by vancluever** [comment]
Thanks @paultyng, yeah, this is how govc does it too (although not opt-in, but rather opt-out, probably due to the much larger risk of running multiple successive/concurrent govc processes versus us here).

I looked at the JSON saved by the SOAP session code and the only sensitive thing leaked is the session ID. So while that of course needs to be treated with care (and is as such mentioned in the docs), it's not as bad as say having full-on credentials there.

Temporary directories _could_ be created ahead of time and passed in, but that's something we can't really control within TF itself. An example is kind of "given" in the acceptance tests on how that can be done, and I'm going to apply a similar pattern to CI so that it uses a path in the working directory so that we can clean it up afterwards.

Will be merging it shortly after making a couple of updates - splitting the defers in the acceptance tests up and also adding a skip for `TF_ACC` as that's what's causing the Travis tests to fail.

**Pull request merged**